`lib/uklibid`: Library identifiers and `SUBBUILD` build system feature

skuenzer opened the pull request:
### Base target

 - Architecture(s): [all]
 - Platform(s): [all]
 - Application(s): [all]


### Additional configuration

 - `CONFIG_LIBUKLIBID=y`

### Description of changes

This PR introduces `lib/uklibid` which is a library that generates at compile-time unique IDs for each library. These IDs are starting from `0` and linearly increase so that they can be used for table lookups with O(1) complexity. `lib/ukdebug` and `lib/ukstore` are first users of this library and adopted accordingly.

As required ground work, this PR also introduces a `SUBBUILD` feature to the build system. This feature adds the ability to place generated files (e.g., pre-processed, compiled) in a subdirectory of the corresponding library build directory instead of just placing them in the top-level directory. For this purpose, the build system looks for a file-scoped variable with the suffix `_SUBBUILD`. Such a variable can contain a path that is created below the library build directory.

Example (`Makefile.uk`):
```Makefile
# Create `myheader.h` under `build/libmylib/include/uk/`
LIBMYLIB_SRCS-y += $(LIBMYLIB_BASE)/myheader.m4>.h
LIBMYLIB_MYHEADER_SUBBUILD += include/uk
```

skuenzer committed:
lib/syscall_shim: Utilize `SUBBUILD` feature within `Makefile.uk`

This commit utilizes the recently added `SUBBUILD` feature of the build
system to natively generate all dynamic headers of the library. This has
several reasons:
 1) The `Makefile.uk` of `syscall_shim` becomes more readable
 2) No custom rule for creating the subdirectory structure is needed
 3) No custom `awk`-rules are needed for generating the headers
 4) The build system can take care of handling build order dependencies

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukdebug: Use lib/uklibid for library names

Every invocation of any of the provided print function macros by
`lib/ukdebug` (`uk_printk()`, `uk_printd()`), caused placing a C-string
containing the library name in the sources. This commit replaces this
C-string with a library identifier which is just an unsigned integer. The
ID is only resolved into a name in the actual print format function
(`_vprint()`). The resolution is done by `uk_libname()` that is using the
ID for a table lookup (O(1)).

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uklibid: Add library skeleton

libuklibid will provide generated identifiers for each selected library.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uklibid: Generate `libraries.in`

`libraries.in` contains a list of currently selected libraries.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uklibid: Compile-time mapping of library names to IDs

This commit introduces macros for mapping a library name to a compile-time
unique identifier with the pre-processor. The library IDs are starting from
`0` and increase linearly. The purpose is that such an ID can be directly
used as index for looking up a corresponding entry within an array.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uklibid: Run-time name resolution of library IDs

This commit introduces the function `uk_libname()` that resolves a library
ID to its library name at run-time. The function returns a reference to a
corresponding C-string with a lookup table.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uklibid: Run-time mapping of library name to identifier

This commit introduces the function `uk_libid()` that finds the library ID
for a given library name at run-time.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build/Makefile.rules: Improve parameter description of functions

Most of our Makefile commands have a preceding usage description in the
form of:
 # func_name $param_desc,$param_desc,...
Because most of these functions reference the parameters by their
position, this commits changes the description to:
 # func_name $1:param_desc,$2:param_desc,...
This should simplify the reading of Makefile function definitions.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build: Only first-level subdirectories with `mk_sub_build_dir`

We have two pairs of functions that are internally used to create the
directory structure the build output directory (normally `build/`):
- The function `mk_sub_build_dir` is intended to create only the first
  level subdirectory under the build output directory for a library.
  `sub_build_dir` is a companion function that returns the full path
  of such a first-level subdirectory for a given library name.
- The function `mk_sub_libbuild_dir` is intended to create further
  subdirectories under a library build directory that was created by
  `mk_sub_build_dir`. This function allows multiple levels of
  subdirectories (example: to create an header include folder:
  'include/uk/bits').
  `sub_libbuild_dir` is the companion function that returns the
  full path of such a library subdirectory.

This commit corrects the implementation so that this behavior is
enforced.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build: Introduce `*_SUBBUILD` for compilation units

This commit adds to the build system the ability to place generated
files (e.g., pre-processed, compiled) in a subdirectory of the
corresponding library build directory (instead of just placing them in the
top-level directory). For this purpose, the build system looks for a
file-scoped variable with the suffix `_SUBBUILD`. Such a variable can
contain a path that is created below the library build directory.

Example (`Makefile.uk`):
 # Create `myheader.h` under `build/libmylib/include/uk/`
 LIBMYLIB_SRCS-y += $(LIBMYLIB_BASE)/myheader.m4>.h
 LIBMYLIB_MYHEADER_SUBBUILD += include/uk

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukstore: Re-use `libraries.in` from `lib/uklibid`

Instead of generating own library identifiers, this commit adopts
`lib/ukstore` to use the same library catalog from `lib/uklibid`. At the
same time, this commit adopt `Makefile.uk` to make use of the `SUBBUILD`
feature of the build system. This is done for simplifying Makefile.uk and
to let the build system natively generate each dynamic headers of
`lib/ukstore`.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukstore: Use `lib/uklibid` for library ID mapping

Instead of implementing own library ID mapping macros and functions, this
commit adopts `lib/ukstore` to use `lib/uklibid` for this purpose.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

michpappas commented on support/build/Makefile.rules:
```diff
@@ -59,8 +59,8 @@ endef
 define mk_sub_build_dir =
-$(if $(shell mkdir -p "$(BUILD_DIR)/$(1)" && cd "$(BUILD_DIR)/$(1)" >/dev/null && pwd),,\
-     $(error could not create directory "$(BUILD_DIR)/$(1)"))
+$(if $(shell mkdir -p "$(BUILD_DIR)/$(notdir $(1))" && cd "$(BUILD_DIR)/$(notdir $(1))" >/dev/null && pwd),,\
```
If `addlib` and friends expect a name, would make more sense to return an error here if someone passes a path instead of silently handling with `notdir`?

skuenzer replied:
I'd prefer avoiding this because for achieving this we would need to do more costly string parsings with Make.

michpappas commented on support/build/Makefile.rules:
```diff
@@ -69,13 +69,13 @@ sub_build_dir = $(addprefix $(BUILD_DIR)/,$(notdir $(1)))
 
 # converts a list of library names to paths pointing to their corresponding object library file
-# libname2olib $libname
+# libname2olib $1:libname
```
`mk_sub_build_dir` above could be documented too

skuenzer replied:
True, it seems I missed that. I'll add it.

michpappas commented on support/build/Makefile.rules:
```diff
@@ -830,28 +850,26 @@ endef
-   $($(call vprefix_srcv,$(1),$(2),$(4),PREPFLAGS))\
-   $($(call vprefix_srcv,$(1),$(2),$(4),PREPFLAGS-y))\
-  ,)\
+$(eval subbuild=$($(call vprefix_src,$(1),$(2),$(4),SUBBUILD)))
```
At first I found the `SUBBUILD` name a bit confusing as it implies a _sub-build_ (ie a build within a build) rather than _a sub-dir under `build/`_, so something along the lines of `BUILD_SUBDIR` would had been be more clear. On the other hand I see that `mk_sub_build_dir` exists for some time now, and to be consistent you would have to also update the rest of `Makefile.rules`...

skuenzer replied:
I went through the same thoughts and chose this name because it was there already (also my mistake) and I was able to keep the variable suffix short. I wanted to avoid anything that has an underscore because resolving uses a lot of underscores already: 
`<LIBNAME>_<SOURCENAME>_<VARIANTNAME>_SUBBUILD`
The thing is that the variable suffix (`_SUBBUILD`) will be used in `Makefile.uk`'s in the future while it is unlikely that somebody uses `mk_sub_build_dir` directly (although there might be exceptions). We can still choose a better name. I also had (something with) `TARGET` in mind... :thinking:

michpappas commented on lib/uklibid/include/uk/libid.h:
```diff
@@ -0,0 +1,9 @@
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+#ifndef LIBUKLIBID_H_
```
I suspect the motivation for not using the usual format `__LIBNAME_H__` is coding conventions? If so, using leading underscores in include guards is actually permitted.

skuenzer replied:
Time for me to study the conventions 😄. I can go back to the (previous) scheme we used with the other libs: `__LIBUKLIBID_H__`.

michpappas commented on lib/uklibid/include/uk/libid.h:
```diff
@@ -6,4 +6,53 @@
+ *          make this pre-processor macro fail.
+ *
+ * @param libname Non-quoted name of the library (no C-string!)
+ * @return Library identifier (unsigned int)
```
Intuitively speaking, would a fixed type like `__u64` make more sense for identifiers?

skuenzer replied:
I was thinking also about this, but rather `__u32` or even `__u16`. Then I thought, we likely have less than `__U16_MAX` libraries in a build (16K libraries...). Then I thought, let the compiler decide an optimal default for the architecture because the values will be at most in the hundreds. A theoretical 8bit target architecture for Unikraft could still support 255 libraries. On the other hand, I agree with you too, using a fixed size integers better communicate size assumptions to calling code. The history tells us that space rockets can explode because calling code may misuse functions simply because not being aware of their original integer size assumptions. A fixed-size integer is probably also better for the case when we need to reserve space for writing libid's out somewhere... in short: you changed my mind about `unsigned int` 😄; I'll change it to `__u16`.

michpappas commented on lib/uklibid/include/uk/libid.h:
```diff
@@ -6,4 +6,53 @@
+#define uk_libid_static(libname) \
+	_uk_libid_static(libname)
+
+#ifndef __LIBNAME__
```
What is the use-case of this?

skuenzer replied:
Pre-processor: This allows `libname` to be resolved when an alias (via macro or macro function) is handed over as argument

michpappas commented on lib/uklibid/include/uk/libid.h:
```diff
@@ -6,4 +6,53 @@
+
+/**
+ * Compile-time mapping of a library name to its unique library identifier
+ * WARNING: Any unknown library name (e.g., library not part of build) will
```
Is it possible to have a user of this header that is not part of the build?

skuenzer replied:
Unikraft's build system should be restrictive enough that this should not be possible. This was a try to document assumptions that were taken during development just for the case we suddenly export headers from Unikraft to externally handled builds.

michpappas commented on lib/uklibid/libname.c:
```diff
@@ -16,3 +19,21 @@ const char *uk_libname(unsigned int libid)
+{
+	unsigned int libid;
+
+	if (unlikely(!libname))
```
An assertion here would avoid an unnecessary conditional, even if it's an optimized one

skuenzer replied:
actually, right :+1:. There is no good reason to call: `uk_libid(NULL)`

michpappas commented on lib/uklibid/include/uk/libid.h:
```diff
@@ -51,6 +51,25 @@ extern "C" {
+ * @return Reference to a '\0'-terminated C-string that contains the library
+ *         name
+ */
+#define uk_libname_self(libid) \
```
Looks like the `libid` parameter creeped in here.

skuenzer replied:
It did... nasty little parameter... 😄

michpappas commented on lib/ukstore/Makefile.uk:
```diff
@@ -1,53 +1,19 @@
+CXXINCLUDES-$(CONFIG_LIBUKSTORE) += -I$(LIBUKSTORE_BUILD)/include
+
+# Cross-reference libraries catalog generated by lib/uklibid
+LIBUKSTORE_LIBRARIES_IN = $(call sub_build_dir,libuklibid)/libraries.in
```
Overall this change is good and cleans up a lot of clutter, one thing that worries me slightly is whether `libraries.in` is a part of `uklibid`'s inner workings, ie considered private. On the other hand I can't think of a better alternative to solve this problem right now.

skuenzer replied:
I agree with you one hundred percent. It is probably a good idea to just document this as bad practice but still best way for this case. I will add such a comment.

michpappas reviewed (changes requested):
Hi @skuenzer nice PR! 🚀 Just a couple of minor issues and some clarifications from my side.

michpappas reviewed (approved):
Looks good, thanks! @razvand it's on you now.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

kubanrob commented on lib/uklibid/Makefile.uk:
```diff
@@ -3,3 +3,23 @@ $(eval $(call addlib_s,libuklibid,$(CONFIG_LIBUKLIBID)))
+		     $(L) \
+		     )) \
+		     $(UK_LIBS) $(UK_LIBS-y) \
+		| tr ' ' '\n' > $@)
```
It would be good if this list would be the stable for the same library names and the exactly same config for the same application.

This can be useful for different reasons:
- if the order of the list changes, compiled olibs need to be recompiled
- following from that: It is impossible to distribute precompiled olibs if they use ukdebug internally
- the ID could be used for logging or tracing
- in general, the build becomes more reproducible

For that reasons, it would be good if there would be either:
- a fixed order of this list (eg. alphabetically)
- known dependencies that need to be met to make the order deterministic, for example: _Make sure that application Makefile includes the libraries in the same order (`L` variable)._

skuenzer replied:
This makes sense, I added a sorting step to ensure determinism. :+1:

kubanrob commented:
I see the same error as `integration / libc-test` on arm64: `E  tar: /github/workspace/_helloworld/.unikraft/build/libmusl/origin: Cannot open: No such file or directory`.

kubanrob commented on lib/ukdebug/outf.h:
```diff
@@ -56,7 +56,7 @@ struct out_dev {
 		struct {
 			int lvl; /* OUTDEV_KERN only */
-			const char *libname;
+			__u16 libid;
```
When building along with lib-musl on arm64: `error: unknown type name ‘__u16’`

skuenzer replied:
Good point, will fix. An include is then only missing.

skuenzer committed:
lib/uklibid: Supporting pre-built libraries

This commit introduces a clearer separation between runtime and
compile-time resolution of library identifiers with the API. This is done
so that libraries can potentially be built off-tree and shipped in binary
form. Each library's own identifier (`uk_libid_self()`) is stored in a
constant variable that is resolved during link-time. The owner of this
variable is `uklibid`, so the identifier is external to a binary library.

A binary delivery of a library can be created with Unikraft (for example:
`make libbinary`) and the resulting library object (likely under
`build/libbinary.o`) can be delivered with a `Config.uk` like:
```KConfig
config LIBBINARY
	bool "Binary example library"
	default n
```

and a `Makefile.uk` like:
```Makefile
UK_OLIBS-$(CONFIG_LIBBINARY) += $(IMPORT_BASE)/libbinary.o
UK_LIBID_EXTRA-$(CONFIG_LIBBINARY) += libbinary
```

Please note that the library must be registered directly in `UK_OLIBS-y`;
`addlib` or `addlib_s` cannot be used in such cases.
Please also note that such a library should restrict dependencies on other
libraries and their configuration options. In particular, the use of macros
and inline functions must be handled with care.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer commented:
> I see the same error as `integration / libc-test` on arm64: `E tar: /github/workspace/_helloworld/.unikraft/build/libmusl/origin: Cannot open: No such file or directory`.

Right, sorry, I applied the fix in the latest version that I just pushed.

nderjung commented:
Hi, please see the builds, they have all failed on:

```
No rule to make target '/github/workspace/lib/uklibid/selfid.c', needed by '/github/workspace/_helloworld/.unikraft/build/libkvmplat/selfid.uklibid.o'
```

skuenzer commented:
> Hi, please see the builds, they have all failed on:
> 
> ```
> No rule to make target '/github/workspace/lib/uklibid/selfid.c', needed by '/github/workspace/_helloworld/.unikraft/build/libkvmplat/selfid.uklibid.o'
> ```

It's now there 😇

skuenzer committed:
build: Support generating .uk files with preprocessor

This commit registers the `.uk` file extension as known extension that can
be generated by the build system with a pre-processor (e.g., M4, AWK).

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build: Provide $(IMPORT_BASE) to `Makefile.uk` files

This commit makes the variable `$(IMPORT_BASE)` usable in `Makefile.uk`
files. This is needed to include libraries that do not use the `addlib`
or `addlib_s` registration functions. An example are libraries that are
shipped in binary form.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer commented:
As we discussed offline, this release now includes support for shipping precompiled libraries, although `uklibid` identifiers are used. For a more detailed explanation, see commit 948604f52b2c993d359b59bd9933cda9821ef325.

kubanrob reviewed (approved):
Looks good to me, thank you @skuenzer!

Reviewed-by: Robert Kuban <robert.kuban@opensynergy.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
